add role argument spec for os, ssh, mysql (#687)
dev-sec/ansible-collection-hardening: roles/ssh_hardening/README.md

# dev-sec.ssh_hardening

![devsec.ssh_hardening](https://github.com/dev-sec/ansible-collection-hardening/workflows/devsec.ssh_hardening/badge.svg)

## Description

This role provides secure ssh-client and ssh-server configurations. It is intended to be compliant with the [DevSec SSH Baseline](https://github.com/dev-sec/ssh-baseline).

Warning: This role disables root-login on the target server! Please make sure you have another user with su or sudo permissions that can login into the server.

<!-- BEGIN_ANSIBLE_DOCS -->
## Supported Operating Systems
| Platform | Versions |
| -------- | -------- |
| EL | 7, 8, 9 |
| Ubuntu | bionic, focal, jammy |
| Debian | buster, bullseye |
| Amazon |  |
| Fedora |  |
| ArchLinux |  |
| SmartOS |  |
| FreeBSD | 12.2 |
| OpenBSD | 7.0 |

## Role Variables

* `network_ipv6_enable`
  * Default: `true`

>   * Description: false if IPv6 is not needed. `ssh_listen_to` must also be set to listen to IPv6 addresses (for example `[::]`).

Reviewer: ```suggestion
  * Description: `false` if IPv6 is not needed. `ssh_listen_to` must also be set to listen to IPv6 addresses (for example `[::]`).
```

  * Default: `/etc/ssh/sshd_config`
  * Description: path of the ssh server configuration file, e.g. `/etc/ssh/sshd_config.d/custom.conf`.
  * Type: str
  * Required: no
* `ssh_server_ports`
  * Default: `["22"]`
  * Description: ports on which ssh-server should listen.
  * Type: list
  * Required: no
* `ssh_client_port`
  * Default: `22`
  * Description: Specifies the port number to connect on the remote host.
  * Type: str
  * Required: no
* `ssh_listen_to`
  * Default: `["0.0.0.0"]`
  * Description: one or more ip addresses, to which ssh-server should listen to. Default i all IPv4 addresses, but should be configured to specific addresses for security reasons
  * Type: list
  * Required: no
* `ssh_host_key_files`
  * Default: ``
  * Description: Host keys for sshd. If empty ['/etc/ssh/ssh_host_rsa_key', '/etc/ssh/ssh_host_ecdsa_key', '/etc/ssh/ssh_host_ed25519_key'] will be used, as far as supported by the installed sshd version.
  * Type: list
  * Required: no
* `ssh_host_rsa_key_size`
  * Default: `4096`
  * Description: Specifies the number of bits in the private host RSA key to create.
  * Type: str
  * Required: no
* `ssh_host_key_algorithms`
  * Default: ``

>   * Description: Host key algorithms that the server offers. If empty the default lst will be used. Otherwise overrides the setting with specified list of algorithms. Check `man sshd_config`, `ssh -Q HostKeyAlgorithms` or other sources for supported algorithms - make sure you check the correct version

Reviewer: ```suggestion
  * Description: Host key algorithms that the server offers. If empty the default list will be used. Otherwise overrides the setting with specified list of algorithms. Check `man sshd_config`, `ssh -Q HostKeyAlgorithms` or other sources for supported algorithms - make sure you check the correct version
```

  * Type: list
  * Required: no
* `ssh_client_host_key_algorithms`
  * Default: ``
  * Description: Specifies the host key algorithms that the client wants to use in order of preference. If empty the default lst will be used. Otherwise overrides the setting with specified list of algorithms. Check `man ssh_config`, `ssh -Q HostKeyAlgorithms` or other sources for supported algorithms - make sure you check the correct version!
  * Type: list
  * Required: no
* `ssh_client_alive_interval`
  * Default: `600`
  * Description: specifies an interval for sending keepalive messages.
  * Type: str
  * Required: no
* `ssh_client_alive_count`
  * Default: `3`
  * Description: Defines the number of acceptable unanswered client alive messages before disconnecting clients.
  * Type: str
  * Required: no
* `ssh_permit_tunnel`
  * Default: `false`
  * Description: true if SSH Port Tunneling is required.
  * Type: bool
  * Required: no
* `ssh_remote_hosts`
  * Default: ``

>   * Description: one or more hosts and their custom options for the ssh-client. Default i empty. See examples in `defaults/ain.yml`

Reviewer: ```suggestion
  * Description: one or more hosts and their custom options for the ssh-client. Default is empty. See examples in `defaults/main.yml`
```

  * Type: list
  * Required: no
* `ssh_permit_root_login`
  * Default: `no`
  * Description: Disable root-login. Set to `'without-password'` or `'yes'` to enable root-login - The quotes are required!
  * Type: str
  * Required: no
* `ssh_allow_tcp_forwarding`
  * Default: `no`
  * Description: 'no' or `False` to disable TCP Forwarding. Set to `'yes'` or `True` to allow TCP Forwarding. If you are using OpenSSH >= 6.2 version, you can specify `'yes'`, `'no'`, `'all'`, `'local'`or`'remote'`.
  * Type: str
  * Required: no
* `ssh_gateway_ports`
  * Default: `false`
  * Description: `false` to disable binding forwarded ports to non-loopback addresses. Set to `true` to force binding on wildcard address. Set to `clientspecified` to allow the client to specify which address to bind to.
  * Type: bool
  * Required: no
* `ssh_allow_agent_forwarding`
  * Default: `false`

>   * Description: false to disable Agent Forwarding. Set to true to allow Agent Forwarding.

Reviewer: this and ff: wo we always mark `true` and `false` as code?

This role uses the default port 22 or the port configured in the inventory to connect to the server. If the default `ssh` port is changed via `ssh_server_ports`, once the ssh server is restarted, it will still try to connect using the previous port. In order to run this role again on the same server the inventory will have to be updated to use the new ssh port.

If idempotency is important, please consider using role [`ssh-hardening-fallback`](https://github.com/nununo/ansible-ssh-hardening-fallback), which is a wrapper around this role that falls back to port 22 if the configured port is unreachable.

## License
Apache License 2.0

## Author and Project Information
Sebastian Gumprich @ Hardening Framework Team

> 

Reviewer: should this be in all README.md files?

Author: Rather on the contrary - it should only be in the collection-readme. I removed it.
